src/chk/chk_engine.c

```diff
@@ -231,24 +231,602 @@ chk_engine_setup_pools(struct chk_instance *ins, bool svc)
+			case PO_COMP_ST_DOWN:
+				down = true;
+				break;
+			case PO_COMP_ST_DOWNOUT:
+				/* Do nothing. */
+				break;
+			case PO_COMP_ST_NEW:
+				/* Dangling new entry in pool map, directly mark as 'DOWNOUT'. */
+				rc = chk_engine_pm_dangling(cpr, t_comp, version,
+							    PO_COMP_ST_DOWNOUT);
+				break;
+			default:
+				D_WARN(DF_ENGINE" hit knownn pool target status %u for "DF_UUIDF
+				       " with rank %u, index %u, ID %u\n", DP_ENGINE(cpr->cpr_ins),
+				       t_comp->co_status, DP_UUID(cpr->cpr_uuid),
+				       t_comp->co_rank, t_comp->co_index, t_comp->co_id);
+				/* Fall through. */
+			case PO_COMP_ST_UP:
+				/* Fall through. */
+			case PO_COMP_ST_UPIN:
+				/* Fall through. */
+			case PO_COMP_ST_DRAIN:
+				down = true;
+				/*
+				 * Some data may be on the lost target, mark as 'DOWN' that
+				 * will be handled via rebuild in subsequent process.
+				 */
+				rc = chk_engine_pm_dangling(cpr, t_comp, version, PO_COMP_ST_DOWN);
+				break;
+			}
+
+			if (rc != 0)
+				goto out;
+
+			t_comp->co_flags |= PO_COMPF_CHK_DONE;
+		}
+
+		rc = chk_engine_pm_dangling(cpr, r_comp, version,
+					    down ? PO_COMP_ST_DOWN : PO_COMP_ST_DOWNOUT);
+		if (rc != 0)
+			goto out;
+
+		r_comp->co_flags |= PO_COMPF_CHK_DONE;
+	}
+
+out:
+	return rc;
+}
+
 static void
 chk_engine_pool_ult(void *args)
 {
 	struct chk_pool_rec	*cpr = args;
+	struct chk_instance	*ins = cpr->cpr_ins;
 	struct chk_bookmark	*cbk = &cpr->cpr_bk;
+	struct ds_pool_svc	*svc = NULL;
+	struct pool_map		*map = NULL;
+	uint32_t		 version;
+	int			 i;
 	int			 rc = 0;
+	int			 rc1 = 0;
 
-#if 0
-	/* TBD: Drive the check since phase CHK__CHECK_SCAN_PHASE__CSP_POOL_MBS. */
+	rc = ds_pool_svc_lookup_leader(cpr->cpr_uuid, &svc, NULL);
+	if (rc != 0)
+		/*
+		 * XXX: Before the phase of CHK__CHECK_SCAN_PHASE__CSP_OBJ_SCRUB, the PS
+		 *	leader drives the check on engine. Current one is not PS leader.
+		 */
+		D_GOTO(out, rc = 0);
 
-	while (!cpr->cpr_stop && cpr->cpr_ins->ci_sched_running) {
-		dss_sleep(300);
+	ABT_mutex_lock(cpr->cpr_mutex);
+
+again:
+	if (cpr->cpr_stop || !ins->ci_sched_running) {
+		ABT_mutex_unlock(cpr->cpr_mutex);
+		goto out;
+	}
+
+	if (cpr->cpr_mbs == NULL) {
+		ABT_cond_wait(cpr->cpr_cond, cpr->cpr_mutex);
+		goto again;
 	}
-#endif
+
+	ABT_mutex_unlock(cpr->cpr_mutex);
+
+	rc = ds_pool_svc_load_map(svc, &map);
+	if (rc != 0)
+		goto out;
+
+	version = pool_map_get_version(map);
+
+	for (i = 0; i < cpr->cpr_shard_nr; i++) {
+		rc = chk_engine_pool_mbs_one(cpr, map, &cpr->cpr_mbs[i], &version);
+		if (rc != 0)
+			break;
+	}
+
+	/* Lookup for dangling entry in the pool map. */
+	if (rc == 0)
+		rc = chk_engine_find_dangling_pm(cpr, map, &version);
+
+	if (cpr->cpr_dirty) {
+		cpr->cpr_dirty = 0;
+
+		chk_bk_update_pool(cbk, cpr->cpr_uuid);
+
+		/*
+		 * Flush the pool map to persistent storage (if not under dryrun mode)
+		 * and distribute the pool map to other pool shards.
+		 */
+		rc1 = ds_pool_svc_flush_map(svc, map, version);
```

Comment:
[Question] The map object already includes the version; the version parameter is unnecessary, isn't it? On the other hand, do you think we should pass a version for ds_pool_svc_flush_map to check before writing the new map? For instance,

    read the map: version x
    change the map
    if the old map is not version x
        try reading and changing again
    write the new map

Reply:
> The map object already includes the version; the version parameter is unnecessary, isn't it?
Right, I will drop the redundant parameter.

> On the other hand, do you think we should pass a version for ds_pool_svc_flush_map to check before writing the new map? For instance,
Under check mode, we disabled node eviction and do not allow reintegration. Means that there will be no pool map refresh except the DAOS check logic itself update the pool map. So it is unnecessary to re-check the pool map version before ds_pool_svc_flush_map(). On the other hand, if the ds_pool_svc_flush_map() is called under non-check mode, then the sponsor needs to do as you described.